File: demo_app/app/src/main/java/com/bandyer/demo_android_sdk/notification/FirebaseCompat.java
/*
 * Copyright (C) 2019 Bandyer S.r.l. All Rights Reserved.
 * See LICENSE.txt for licensing information
 */

package com.bandyer.demo_android_sdk.notification;

import android.content.Context;
import android.util.Log;
import android.widget.Toast;

import androidx.annotation.NonNull;

import com.bandyer.demo_android_sdk.utils.networking.ConnectionStatusChangeReceiver;
import com.bandyer.demo_android_sdk.utils.networking.MockedNetwork;
import com.bandyer.demo_android_sdk.utils.networking.OnNetworkConnectionChanged;
import com.bandyer.demo_android_sdk.utils.storage.ConfigurationPrefsManager;
import com.bandyer.demo_android_sdk.utils.storage.LoginManager;
import com.google.firebase.FirebaseApp;
import com.google.firebase.FirebaseOptions;
import com.google.firebase.iid.FirebaseInstanceId;

import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

/**
 * This class is for Bandyer usage only
 * You should implement your own logic for notification handling
 *
 * @author kristiyan
 */
public class FirebaseCompat {

    private static boolean deviceRegistered = false;

    private static OnNetworkConnectionChanged onNetworkConnectionChanged = (context, connected) -> {
        if (deviceRegistered || !connected) return;
        registerDevice(context);
    };

    static void unregisterDevice(Context context, String loggedUser) {
        if (!deviceRegistered) return;
        ConnectionStatusChangeReceiver.unRegister(context);
        deviceRegistered = false;
        try {
            FirebaseInstanceId.getInstance().getInstanceId().addOnSuccessListener(instanceIdResult -> {
                Thread post = new Thread(() -> {
                    String devicePushToken = instanceIdResult.getToken();
                    MockedNetwork.unregisterDeviceForPushNotification(context, loggedUser, devicePushToken);
                    try {
                        FirebaseInstanceId.getInstance().deleteInstanceId();
                        FirebaseApp.getInstance().delete();
                    } catch (Throwable e) {
                        e.printStackTrace();
                    }
                });
                post.start();
            });

        } catch (Throwable e) {
            e.printStackTrace();
        }
    }

    static void registerDevice(Context context) {
        if (deviceRegistered) return;
        ConnectionStatusChangeReceiver.register(context, onNetworkConnectionChanged);
        if (!ConnectionStatusChangeReceiver.isConnected(context)) return;

        refreshConfiguration(context, () -> {
            try {
                FirebaseInstanceId.getInstance().getInstanceId().addOnSuccessListener(instanceIdResult -> {
                    String devicePushToken = instanceIdResult.getToken();
                    MockedNetwork.registerDeviceForPushNotification(context, LoginManager.getLoggedUser(context), devicePushToken, new Callback<Void>() {
                        @Override
                        public void onResponse(@NonNull Call<Void> call, @NonNull Response<Void> response) {
                            if (!response.isSuccessful()) {
                                Log.e("PushNotification", "Failed to register device for push notifications!");
                                return;
                            }
                            deviceRegistered = true;
                        }

                        @Override
                        public void onFailure(@NonNull Call<Void> call, @NonNull Throwable t) {
                            Log.e("PushNotification", "Failed to register device for push notifications!");
                        }
                    });
                }).addOnFailureListener(error -> {
                    Toast.makeText(context, "Wrong configuration for FCM.\nYou will not receive any notifications!", Toast.LENGTH_LONG).show();
                });
            } catch (Throwable e) {
                e.printStackTrace();
            }
        });
    }

    /**
     * This is necessary to change the projectId at runtime
     * In normal use-cases you shall not need to refresh the configuration as it will use the google-services.json file.
     */
    private static void refreshConfiguration(Context context, Runnable onComplete) {
        Thread post = new Thread(() -> {
            try {
                FirebaseApp.getInstance().delete();
            } catch (Throwable e) {
                e.printStackTrace();
            } finally {
                String projectId = ConfigurationPrefsManager.getFirebaseProjectNumber(context);
                FirebaseApp app = FirebaseApp.initializeApp(context, new FirebaseOptions.Builder()
                        .setGcmSenderId(projectId)
                        .setProjectId(projectId)
                        .setApplicationId(context.getPackageName())
                        .build());
                FirebaseInstanceId.getInstance(app).getInstanceId();
                onComplete.run();
            }
        });
        post.start();
    }
}

File: app_utilities/src/main/java/com/bandyer/app_utilities/notification/FirebaseCompat.kt
/*
 * Copyright (C) 2020 Bandyer S.r.l. All Rights Reserved.
 * See LICENSE.txt for licensing information
 */
package com.bandyer.app_utilities.notification

import android.content.Context
import android.util.Log
import android.widget.Toast
import com.bandyer.app_configuration.external_configuration.model.PushProvider
import com.bandyer.app_utilities.networking.ConnectionStatusChangeReceiver.Companion.isConnected
import com.bandyer.app_utilities.networking.ConnectionStatusChangeReceiver.Companion.register
import com.bandyer.app_utilities.networking.ConnectionStatusChangeReceiver.Companion.unRegister
import com.bandyer.app_utilities.networking.MockedNetwork.registerDeviceForPushNotification
import com.bandyer.app_utilities.networking.MockedNetwork.unregisterDeviceForPushNotification
import com.bandyer.app_utilities.networking.OnNetworkConnectionChanged
import com.bandyer.app_utilities.storage.ConfigurationPrefsManager
import com.bandyer.app_utilities.storage.LoginManager
import com.google.android.gms.common.ConnectionResult
import com.google.android.gms.common.GoogleApiAvailability
import com.google.firebase.FirebaseApp
import com.google.firebase.FirebaseOptions
import com.google.firebase.iid.FirebaseInstanceId
import com.google.firebase.iid.InstanceIdResult
import com.google.firebase.messaging.FirebaseMessaging
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response

/**
 * This class is for Bandyer usage only
 * You should implement your own logic for notification handling
 *
 * @author kristiyan
 */
object FirebaseCompat {
    private var deviceRegistered = false
    private val onNetworkConnectionChanged: OnNetworkConnectionChanged = object : OnNetworkConnectionChanged {
        override fun onNetworkConnectionChanged(context: Context, connected: Boolean) {
            if (deviceRegistered || !connected) return
            registerDevice(context, LoginManager.getLoggedUser(context))
        }
    }

    fun unregisterDevice(context: Context?, loggedUser: String?) {
        if (!deviceRegistered) return
        loggedUser ?: return
        unRegister(context!!)
        deviceRegistered = false
        try {
            FirebaseInstanceId.getInstance().instanceId.addOnSuccessListener { instanceIdResult: InstanceIdResult ->
                val post = Thread {
                    val devicePushToken = instanceIdResult.token
                    val configuration = ConfigurationPrefsManager.getConfiguration(context)
                    unregisterDeviceForPushNotification(loggedUser, devicePushToken, configuration.apiKey!!, configuration.appId!!, configuration.environment!!)
                    try {
                        FirebaseInstanceId.getInstance().deleteInstanceId()
                        FirebaseApp.getInstance().delete()
                    } catch (e: Throwable) {
                        e.printStackTrace()
                    }
                }
                post.start()
            }
        } catch (e: Throwable) {
            e.printStackTrace()
        }
    }

    @JvmStatic
    fun registerDevice(context: Context, loggedUser: String) {
        if (deviceRegistered) return
        register(context, onNetworkConnectionChanged)
        if (!isConnected(context)) return
        refreshConfiguration(context, Runnable {
            try {
                FirebaseInstanceId.getInstance().instanceId.addOnSuccessListener { instanceIdResult: InstanceIdResult ->
                    val devicePushToken = instanceIdResult.token
                    val configuration = ConfigurationPrefsManager.getConfiguration(context)
                    registerDeviceForPushNotification(loggedUser, PushProvider.FCM, devicePushToken, configuration.apiKey!!, configuration.appId!!, configuration.environment!!, object : Callback<Void?> {
                        override fun onResponse(call: Call<Void?>, response: Response<Void?>) {
                            if (!response.isSuccessful) {
                                Log.e("PushNotification", "Failed to register device for push notifications!")
                                return@onResponse
                            }
                            deviceRegistered = true
                        }

                        override fun onFailure(call: Call<Void?>, t: Throwable) {
                            Log.e("PushNotification", "Failed to register device for push notifications!")
                        }
                    })
                }.addOnFailureListener { error: Exception? -> Toast.makeText(context, "Wrong configuration for FCM.\nYou will not receive any notifications!", Toast.LENGTH_LONG).show() }
            } catch (e: Throwable) {
                e.printStackTrace()
            }
        })
    }

    /**
     * This is necessary to change the projectId at runtime
     * In normal use-cases you shall not need to refresh the configuration as it will use the google-services.json file.
     */
    fun refreshConfiguration(context: Context, onComplete: Runnable, notifications: Boolean = true) {
        val post = Thread {
            kotlin.runCatching { FirebaseApp.getInstance().delete() }
            kotlin.runCatching {
                val configuration = ConfigurationPrefsManager.getConfiguration(context)
                val app = FirebaseApp.initializeApp(context, FirebaseOptions.Builder()
                        .setGcmSenderId(configuration.projectNumber)
                        .setProjectId(configuration.firebaseProjectId)
                        .setApplicationId(configuration.firebaseMobileAppId!!)
                        .setApiKey(configuration.firebaseApiKey!!)
                        .build())
                FirebaseInstanceId.getInstance(app).instanceId
                if (notifications) FirebaseMessaging.getInstance().isAutoInitEnabled = true
                onComplete.run()
            }.onFailure { it.printStackTrace() }
        }
        post.start()
    }

    fun isGmsAvailable(context: Context?): Boolean {
        var isAvailable = false
        if (null != context) {
            val result = GoogleApiAvailability.getInstance().isGooglePlayServicesAvailable(context)
            isAvailable = ConnectionResult.SUCCESS == result
        }
        Log.i("PushNotification", "isGmsAvailable: $isAvailable")
        return isAvailable
    }
}